pkg/cmd/cli/cmd/set/probe.go

```diff
@@ -0,0 +1,386 @@
+package set
+
+import (
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"strconv"
+
+	"github.com/spf13/cobra"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"k8s.io/kubernetes/pkg/util/intstr"
+	"net"
+	"strings"
+)
+
+const (
+	probeLong = `
+Set or remove a liveness or readiness probe from a pod or pod template
+
+Each container in a pod may define one or more probes that are used for general health
+checking. A liveness probe is checked periodically to ensure the container is still healthy:
+if the probe fails, the container is restarted. Readiness probes set or clear the ready
+flag for each container, which controls whether the container's ports are included in the list
+of endpoints for a service and whether a deployment can proceed. A readiness check should
+indicate when your container is ready to accept incoming traffic or begin handling work.
+Setting both liveness and readiness probes for each container is highly recommended.
+
+The three probe types are:
+
+1. Open a TCP socket on the pod IP
+2. Perform an HTTP GET against a URL on a container that must return 200 OK
+3. Run a command in the container that must return exit code 0
+
+Containers that take a variable amount of time to start should set generous
+initial-delay-seconds values, otherwise as your application evolves you may suddenly begin
+to fail.`
+
+	probeExample = `  # Clear both readiness and liveness probes off all containers
+  $ %[1]s probe dc/registry --unset --readiness --liveness
+
+  # Set an exec action as a liveness probe to run 'echo ok'
+  $ %[1]s probe dc/registry --liveness -- echo ok
+
+  # Set a readiness probe to try to open a TCP socket on 3306
+  $ %[1]s probe rc/mysql --readiness --open-tcp=3306
+
+  # Set an HTTP readiness probe for port 8080 and path /healthz over HTTP on the pod IP
+  $ %[1]s probe dc/webapp --readiness --get-url=http://:8080/healthz
+
+  # Set an HTTP readiness probe over HTTPS on 127.0.0.1 for a hostNetwork pod
+  $ %[1]s probe dc/router --readiness --get-url=https://127.0.0.1:1936/stats
+
+  # Set only the initial-delay-seconds field on all deployments
+  $ %[1]s probe dc --all --readiness --initial-delay-seconds=30`
+)
+
+type ProbeOptions struct {
+	Out io.Writer
+	Err io.Writer
+
+	Filenames         []string
+	ContainerSelector string
+	Selector          string
+	All               bool
+
+	Builder *resource.Builder
+	Infos   []*resource.Info
+
+	Encoder runtime.Encoder
+
+	ShortOutput bool
+	Mapper      meta.RESTMapper
+
+	PrintObject            func(runtime.Object) error
+	UpdatePodSpecForObject func(runtime.Object, func(spec *kapi.PodSpec) error) (bool, error)
```

Comment:
For a single use field the type doesn't make anything better less
understandable (I have to go find the method, add Godoc to it, etc).  Also,
Factory isn't the wrong pattern - this is the simplest way to express a set
of common behaviors that multiple commands use that can be abstracted.
Passing the factory onto Options is definitely the wrong pattern.

On Tue, Feb 16, 2016 at 4:08 PM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/set/probe.go
> <https://github.com/openshift/origin/pull/7331#discussion_r53078585>:
>
> > +
> > +	Filenames         []string
> > +	ContainerSelector string
> > +	Selector          string
> > +	All               bool
> > +
> > +	Builder *resource.Builder
> > +	Infos   []*resource.Info
> > +
> > +	Encoder runtime.Encoder
> > +
> > +	ShortOutput bool
> > +	Mapper      meta.RESTMapper
> > +
> > +	PrintObject            func(runtime.Object) error
> > +	UpdatePodSpecForObject func(runtime.Object, func(spec *kapi.PodSpec) error) (bool, error)
>
> please start typing these or this will become another Factory.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7331/files#r53078585>.
>